### PR TITLE
5.4 Add documentation for slave high availability

### PR DESCRIPTION
**Pull request by bmansheim**

### CODE REVIEW

File: content/rs/administering/database-operations/slave-ha.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+Also, when DB is relocated, the start time is marked in CCS, in BDB key, in field slave_ha_last_relocation_time.
+
+Until DB cooldown has passed, the DB may not be relocated again, from another node. 
+
+DB cooldown should be much longer than cluster cooldown, to prevent a "node killing shard" from moving through cluster nodes, killing them in the process.
+
+Cooldown periods are configurable via rladmin.
+
+Constraints
```
Comment (bmansheim) [inline content/rs/administering/database-operations/slave-ha.md]: @Paz-Yanover Do we need to mention these constraints?

Comment (bmansheim) [reply content/rs/administering/database-operations/slave-ha.md]: All of the constraints of shard relocation apply, such as rack-awareness.

File: content/rs/administering/database-operations/slave-ha.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+that does not also have the master shard.
+    All of the constraints of shard relocation apply, such as rack-awareness.
+1. The data from the master shard is replicated to the new slave shard.
+
+Note: If slave HA is enabled for a database that does not have any slave shards, 
+the failover creates a new master on an available node. The data from the 
+original master is lost but a new master is available for new transactions. This 
+is a typical for caching use cases.
+
+## Grace Period
```
Comment (RoeyPrat) [inline content/rs/administering/database-operations/slave-ha.md]: I'd add that you can see slaveHA config using `rladmin info cluster`

File: content/rs/administering/database-operations/slave-ha.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+By default, slave HA has a 15 minute grace period before new slave shards are created. 
+To configure this grace period from rladmin, run:
+
+    rladmin tune cluster slave_ha_grace_period <time_in_seconds>
+
+## Shard Priority
+
+Slave HA creates new slave shards for databases according to this order of priority:
```
Comment (RoeyPrat) [inline content/rs/administering/database-operations/slave-ha.md]: is it not better to say migrate? i'm asking, not sure.

File: content/rs/administering/database-operations/slave-ha.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+Similarly, the database cannot go through another slave relocation until the 
+cooldown period for the database passes, which is 24 hours.
+
+To configure this grace period from rladmin, run:
+
+    * For a cluster:
+
+        rladmin tune cluster slave_ha_cooldown_period <time_in_seconds>
+    
+    * For a database:
```
Comment (RoeyPrat) [inline content/rs/administering/database-operations/slave-ha.md]: suggestion: "for all databases"

File: content/rs/administering/database-operations/slave-ha.md [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+## Cooldown Periods
+
+Both the cluster and the database have cooldown periods. Slave relocation cannot run 
+on another node in a cluster until the cooldown period for the cluster passes, which is 1 hour. 
+Similarly, the database cannot go through another slave relocation until the 
+cooldown period for the database passes, which is 24 hours.
+
+To configure this grace period from rladmin, run:
+
+    * For a cluster:
```
Comment (RoeyPrat) [inline content/rs/administering/database-operations/slave-ha.md]: suggestion: for *the* cluster

File: content/rs/administering/database-operations/slave-ha.md [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+## Alerts
+
+These alerts are sent during slave HA:
+
+* Shard relocation begins after the grace period
+* Shard relocation fails because there is no available node (Sent hourly)
+* Shard relocation is delayed because of the cooldown period
+
+## Configuring High Availability for Slave Shards
```
Comment (RoeyPrat) [inline content/rs/administering/database-operations/slave-ha.md]: suggestion: push this up to before grace period. it is the TLDR...

File: content/rs/administering/database-operations/slave-ha.md [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+slave relocation creates a new slave shard and replicates the data from the master 
+shard to the new slave shard. For example:
+
+1. Node:2 has a master shard and node:3 has the corresponding the slave shard.
+1. Node:2 fails and triggers a failover.
+1. The slave shard on node:3 is promoted to master.
+1. If slave HA is enabled, a new slave shard is created on an available node that 
+that does not also have the master shard.
+    All of the constraints of shard relocation apply, such as rack-awareness.
+1. The data from the master shard is replicated to the new slave shard.
```
[the remark this answers is not in the document]
Comment (bmansheim) [reply content/rs/administering/database-operations/slave-ha.md]: Hugo (gohugo.io) translates the 1. into ordered lists.

File: content/rs/administering/database-operations/slave-ha.md [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+Note: If slave HA is enabled for a database that does not have any slave shards,
```
[the remark this answers is not in the document]
Comment (RoeyPrat) [reply content/rs/administering/database-operations/slave-ha.md]: > When slave HA is enabled for a non-persistent database
> that does not have any slave shards (for example, a cache database),
I think you start with a clarification of what we mean by cache. AFAIK we never say this explicitly. 
a cache database is a non-replicated non-persistent database. it should only be used in cases where data loss is acceptable when nodes fail.

[the remark this answers is not in the document]
Comment (bmansheim) [reply content/rs/administering/database-operations/slave-ha.md]: Right. These sections were moved to another page to be written for a future release. 5.4 will not mention how slave HA works with caching databases. According to Paz, customers will likely assume that slave HA doesn't apply to databases without slaves.

**Comment by RoeyPrat** [comment]
@yaeltzirulnikov FYI

File: content/rs/administering/database-operations/slave-ha.md [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+1. The data from the master shard is replicated to the new slave shard.
+
+## Configuring High Availability for Slave Shards
+
+You can enable slave HA either for:
+
+* Cluster - All databases in the cluster use slave HA
+* Database - Only the specific database uses slave HA
+
+By default, slave HA is set to disabled at the cluster level. You can enable it
```
Comment (Paz-Yanover) [inline content/rs/administering/database-operations/slave-ha.md]: We should make it clear that if the slave HA is disabled in the cluster level, the slave HA will be disabled for each database in the cluster, even when setting slave HA enabled for a specific database

File: content/rs/administering/database-operations/slave-ha.md [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+### Grace Period
+
+By default, slave HA has a 15-minute grace period after node failure and before new slave shards are created. 
+To configure this grace period from rladmin, run:
+
+    rladmin tune cluster slave_ha_grace_period <time_in_seconds>
+
+### Shard Priority
+
+Slave HA creates new slave shards for databases according to this order of priority:
```
Comment (Paz-Yanover) [inline content/rs/administering/database-operations/slave-ha.md]: We'd better explain that the slave shards migration is performed based on priority as the resources are limited so maybe some migration won't happen due to resources (- memory) limitation.
In addition we need to mention that the order of priority is from the most important to the less important or such (i.e. first item "wins" the second one etc.)

File: content/rs/administering/database-operations/slave-ha.md [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+
+1. slave_ha_priority - You can assign a priority number to a specific database to 
+make sure that its slave shards are created first. To assign priority to a database, run:
+
+    rladmin tune db <bdb_uid> slave_ha_priority <positive integer>
```
Comment (Paz-Yanover) [inline content/rs/administering/database-operations/slave-ha.md]: Higher positive integer has a higher priority or a lower...?

File: content/rs/administering/database-operations/slave-ha.md [hunk 12]
```diff
@@ -0,0 +1,91 @@
+---
+Title: High Availability for Slave Shards
+description: 
+weight: $weight
+alwaysopen: false
+categories: ["RS"]
+---
+When Replication is enabled and master shard fails, the slave shard is automatically promoted to a master shard 
+to maintain data availability. This creates a single point of failure until a new slave shard is manually created.
```
Comment (Paz-Yanover) [inline content/rs/administering/database-operations/slave-ha.md]: What about the situation where the slave shard fails? It's also relevant here. In such case we also have a single point of failure, and then a new slave shard is created and the data is replicated from the master to the new slave shard

File: content/rs/administering/database-operations/slave-ha.md (hunk 11 above)
Comment (bmansheim) [reply content/rs/administering/database-operations/slave-ha.md]: @RoeyPrat Does the higher integer in slave_priority indicate higher or lower priority?

File: content/rs/administering/database-operations/slave-ha.md [hunk 13]
```diff
@@ -61,14 +61,13 @@ make sure that its slave shards are created first. To assign priority to a datab
 
 1. Slave shards of CRDBs - because the sync between the various replicas is critical, 
 and should be performed using slave shards.
-1. Cache databases - because such databases are unavailable if it is not moved to an available node.
```
Comment (bmansheim) [inline content/rs/administering/database-operations/slave-ha.md]: @Paz-Yanover Did you intentionally remove cache databases?

File: content/rs/administering/database-operations/slave-ha.md (hunk 11 above)
Comment (RoeyPrat) [reply content/rs/administering/database-operations/slave-ha.md]: higher priority integer is higher.
0 is no priority (sort by the other parameters)
1 is a priority
2 is a higher priority...

**Comment by bmansheim** [comment]
@Paz-Yanover Please review https://github.com/RedisLabs/redislabs-docs/pull/108/commits/8b95462c16a4afb61f401dadd1e3d0ae7bfbc344

File: content/rs/administering/database-operations/slave-ha.md (hunk 9 above)
Comment (bmansheim) [reply content/rs/administering/database-operations/slave-ha.md]: @Paz-Yanover See https://github.com/RedisLabs/redislabs-docs/pull/108/commits/44bcb4cbb4ad6a287b7d221fdc10721b6eb0eec7

Comment (Paz-Yanover) [reply content/rs/administering/database-operations/slave-ha.md]: Approved. Well done :)

File: content/rs/administering/database-operations/slave-ha.md (hunk 13 above)
Comment (Paz-Yanover) [reply content/rs/administering/database-operations/slave-ha.md]: Yes

File: content/rs/administering/database-operations/slave-ha.md (hunk 12 above)
Comment (bmansheim) [reply content/rs/administering/database-operations/slave-ha.md]: See https://github.com/RedisLabs/redislabs-docs/pull/108/commits/86ab4797cf30795976cc988463a416b0d5f04b03

**Pull request merged**